Source/bmalloc/libpas/src/libpas/pas_large_heap.c

```
@@ -155,6 +159,52 @@ static pas_allocation_result allocate_impl(pas_large_heap* heap,
     return result;
 }
 
+static pas_allocation_result delegated_allocate_impl(size_t size,
+                                                     size_t alignment,
+                                                     pas_allocation_mode allocation_mode,
+                                                     const pas_heap_config* heap_config)
+{
+    pas_allocation_result result;
+
+    PAS_TESTING_ASSERT(pas_system_heap_is_enabled(heap_config->kind));
+    switch (allocation_mode) {
+    case pas_non_compact_allocation_mode:
+        result.begin = (uintptr_t)pas_system_heap_memalign(alignment ? alignment : 1, size);
+        break;
+    case pas_always_compact_allocation_mode:
+        result.begin = (uintptr_t)pas_system_heap_memalign_compact(alignment ? alignment : 1, size);
+        break;
+    case pas_maybe_compact_allocation_mode:
+        PAS_ASSERT_NOT_REACHED();
+        break;
+    }
+
+    if (!result.begin) {
+        result.did_succeed = false;
+        return result;
+    }
+    result.zero_mode = pas_zero_mode_may_have_non_zero;
+    result.did_succeed = true;
+
+    return result;
+}
+
+PAS_IGNORE_WARNINGS_BEGIN("unreachable-code")
+static bool should_delegate_user_allocation_to_system_malloc(size_t size,
+                                                             pas_allocation_mode allocation_mode,
+                                                             const pas_heap_config* heap_config)
+{
+    if (!PAS_MTE_IS_HARDENED)
+        return false;
+    if (size < PAS_MAX_MTE_TAGGABLE_OBJECT_SIZE)
+        return false;
+    return (pas_system_heap_is_enabled(heap_config->kind)
+            && heap_config->delegate_large_user_allocations
+            && heap_config->mmap_capability == pas_may_mmap
```

This combination seems like it's easy to create configs that silently ignores `delegate_large_user_allocations`. Maybe we should instead disallow that config combination?

Also, why is `pas_may_mmap` required for delegation?

---

That's a good point, I'll find a place to check for and exclude that combo.

> why is pas_may_mmap required for delegation?

Because when it's set to false, that means that the user passed already-mapped memory into libpas for the allocator to use -- e.g. the JIT allocator. In that case the user depends on the memory being in a certain region, so we can't go and allocate out of who-knows-where instead.